apps/prairielearn/src/pages/publicQuestionPreview/publicQuestionPreview.ts

```diff
@@ -0,0 +1,109 @@
+import { selectQuestion } from '../../models/question';
+import { selectCourse } from '../../models/course';
+import { processSubmission } from '../../lib/questionPreview';
+import { IdSchema, UserSchema } from '../../lib/db-types';
+import { z } from 'zod';
+
+import ERR = require('async-stacktrace');
+import { Router } from 'express';
+import async = require('async');
+import path = require('path');
+import error = require('@prairielearn/error');
+import question = require('../../lib/question');
+import LogPageView = require('../../middlewares/logPageView');
+const logPageView = LogPageView(path.basename(__filename, '.ts'));
+
+const router = Router({ mergeParams: true });
+
+async function setLocals(req, res) {
+  res.locals.user = UserSchema.parse(res.locals.authn_user);
+  res.locals.authz_data = { user: res.locals.user };
+  res.locals.course = await selectCourse({ course_id: req.params.course_id });
+  res.locals.question = await selectQuestion({ question_id: req.params.question_id });
+  if (!res.locals.question.shared_publicly) {
+    throw error.make(404, 'Not Found');
+  }
+  return;
+}
+
+router.post('/', function (req, res, next) {
+  setLocals(req, res)
+    .then(() => {
+      if (req.body.__action === 'grade' || req.body.__action === 'save') {
+        processSubmission(req, res, function (err, variant_id) {
+          if (ERR(err, next)) return;
+          res.redirect(
+            res.locals.urlPrefix +
+              '/question/' +
+              res.locals.question.id +
+              '/preview/?variant_id=' +
+              variant_id,
+          );
+        });
+      } else if (req.body.__action === 'report_issue') {
+        // we don't care about reporting issues for public facing previews
```

Comment:
You should at the very least redirect back to the existing page, but consider explicitly propagating an error. If you do nothing here, request processing will hang and we'll be stuck with a connection that won't be closed until the browser times out.

Reply:
propagating to where exactly? Do you want it to actually be shown to the course that owns the question? If so, we probably need some new UI that shows it as being from a public preview, not from a student. That seems reasonable but probably beyond the scope of this PR.

Reply:
this seems like it could potentially be confusing to a user if it showed up in production by accident. Maybe I should just stick with a self-redirect?

Reply:
I just did self-redirect.

Reply:
I also edited the comment to make it clear that we don't 'currently' report these issues (implying we may do something more at some future point).